src/Field.sol

```diff
@@ -0,0 +1,172 @@
+
+                let result_ptr := mload(0x40)
+                if iszero(staticcall(not(0), 0x05, pointer, 0xc0, result_ptr, 0x20)) {
+                    revert(0, 0)
+                }
+                answer := mload(result_ptr)
+            }
+
+            // This calculates the the Legendre symbol of base modulo modulus. Again, it is assumed modulus is prime.
+            function is_square(base, modulus) -> isSquare
+            {
+                let exponent := div(sub(modulus, 1),2)
+                let exp_result := pow_mod(base, exponent, modulus)
+                isSquare := iszero(sub(exp_result, 1))
+            }
+
+            function require(condition) {
+                if iszero(condition) { revert(0, 0) }
+            }
+
+            require(is_square(_x, _mod))
+
+            if iszero(sub(mod(_mod, 4),3)) {
+                let exponent := div(add(_mod, 1), 4)
+                let answer_ptr := mload(0x40)
+                mstore(answer_ptr, pow_mod(_x, exponent,_mod))
+                return(answer_ptr, 0x20)
+            }
+
+            let s := sub(_mod, 1)
+            let e := 0
+            for { } iszero(mod(s, 2)) { } {
+                s := div(s, 2)
+                e := add(e, 1)
+            }
+
+            let n := 2
+            for { } is_square(n, _mod) {} {
+                n := add(n, 1)
+            }
+
+            let x := pow_mod(_x, div(add(s, 1), 2), _mod)
+            let b := pow_mod(_x, s, _mod)
+            let g := pow_mod(n, s, _mod)
+            let r := e
+
+            for { } 1 { } {
+                let t := b
+                let m := 0
+                for { } lt(m, r) { m := add(m, 1) } {
+                    if eq(t, 1) {
+                        break
+                    }
+                    t := mulmod(t, t, _mod)
+                }
+
+                if iszero(m) {
+                    let return_ptr := mload(0x40)
+                    mstore(return_ptr, x)
+                    return(return_ptr, 0x20)
+                }
+
+                let gs := pow_mod(g, exp(2, sub(r, add(m, 1))), _mod)
+                g := mulmod(gs, gs, _mod)
+                x := mulmod(x, gs, _mod)
+                b := mulmod(b, g, _mod)
+                r := m
+            }
+        }
+        revert();
+    }
+
+    // Returns the 4 limbs in little-endian order of a 256-bit field element.
+    function extractLimbs(uint256 x) public pure returns (uint256, uint256, uint256, uint256) {
+            uint256 limb1 = (0x000000000000000000000000000000000000000000000000ffffffffffffffff & x);
+            uint256 limb2 = (0x00000000000000000000000000000000ffffffffffffffff0000000000000000 & x) >> 64;
+            uint256 limb3 = (0x0000000000000000ffffffffffffffff00000000000000000000000000000000 & x) >> 128;
+            uint256 limb4 = (0xffffffffffffffff000000000000000000000000000000000000000000000000 & x) >> 192;
+
+            return (limb1, limb2, limb3, limb4);
+    }
```

Comment:
If this passes tests, this should work, but do you have any intuition for why returning 4x u256 works better than 4x u64 here?